File: tests/e2e/installed/typescript/contract_probe.mjs
import assert from "node:assert/strict";
import { createAgent } from "@microsoft/amplifier-agent";

const probe = JSON.parse(process.env.E2E_CONTRACT_PROBE);
const effects = [];
const options = {
  tools: probe.tool ? [{
    name: "observe", description: "Observe",
    inputSchema: { $schema: "https://json-schema.org/draft/2020-12/schema", type: "object" },
    handler: async (_arguments, context) => {
      effects.push(context.call_id);
      return "Confirmed effect";
    },
  }] : [],
  approvals: "allow",
};
let agent;
try { agent = await createAgent(options); }
catch (error) {
  assert.equal(probe.mode, "invalid_host");
  console.log(JSON.stringify({ kind: "error", error: {
    code: error.code, category: error.category, message: error.message,
    remedy: error.remedy, retryable: error.retryable,
  } }));
}
if (agent) {
  assert.notEqual(probe.mode, "invalid_host");
  const events = [], histories = [];
  try {
    let session = await agent.createSession({ sessionId: "installed-contract" });
    for (let index = 0; index < probe.rounds; index++) {
      if (index && probe.mode === "restart") {
        await session.close();
        await agent.close();
        agent = await createAgent(options);
        session = await agent.resumeSession("installed-contract");
      }
      const turn = await session.startTurn({ content: [{ type: "text", text: `Visible question ${index}` }] });
      const observed = [];
      for await (const event of turn.events()) observed.push(event);
      assert.deepEqual(observed.map((event) => event.sequence), observed.map((_, i) => BigInt(i + 1)));
      assert.equal(observed[0].type, "turn_started");
      assert.equal(observed.at(-1).type, "terminal");
      const result = observed.at(-1).payload;
      assert.equal(result.state, "success", JSON.stringify(result.error));
      assert.equal(result.content.map((part) => part.text).join(""), "Wire reply");
      assert.deepEqual(session.history.at(-1).result, result);
      events.push(observed.map((event) => event.type));
      histories.push(session.history.length);
    }
    await session.close();
  } finally { await agent.close(); }
  console.log(JSON.stringify({ kind: "done", events, histories, effects }));
}

File: tests/e2e/installed/python/contract_probe.py
"""Observe installed Python handles without importing the producer checkout."""

import asyncio
import json
import os

from amplifier_agent import (
    AgentError,
    AgentOptions,
    SessionOptions,
    TextPart,
    Tool,
    TurnInput,
    create_agent,
)


async def main():
    probe = json.loads(os.environ["E2E_CONTRACT_PROBE"])
    effects = []

    async def effect(arguments, context):
        effects.append(context.call_id)
        return "Confirmed effect"

    options = AgentOptions(
        tools=[
            Tool(
                "observe",
                "Observe",
                {"$schema": "https://json-schema.org/draft/2020-12/schema", "type": "object"},
                effect,
            )
        ]
        if probe.get("tool")
        else [],
        approvals="allow",
    )
    try:
        agent = await create_agent(options)
    except AgentError as error:
        assert probe["mode"] == "invalid_host"
        print(json.dumps({"kind": "error", "error": vars(error)}), flush=True)
        return
    assert probe["mode"] != "invalid_host"
    events, histories = [], []
    try:
        session = await agent.create_session(SessionOptions(session_id="installed-contract"))
        for index in range(probe["rounds"]):
            if index and probe["mode"] == "restart":
                await session.close()
                await agent.close()
                agent = await create_agent(options)
                session = await agent.resume_session("installed-contract")
            turn = await session.start_turn(TurnInput([TextPart(f"Visible question {index}")]))
            observed = [event async for event in turn.events()]
            assert [event.sequence for event in observed] == list(range(1, len(observed) + 1))
            assert observed[0].type == "turn_started" and observed[-1].type == "terminal"
            result = observed[-1].payload
            assert result.state == "success", vars(result.error) if result.error else result
            assert "".join(part.text for part in result.content) == "Wire reply"
            assert session.history[-1].result == result
            events.append([event.type for event in observed])
            histories.append(len(session.history))
        await session.close()
    finally:
        await agent.close()
    print(
        json.dumps({"kind": "done", "events": events, "histories": histories, "effects": effects}),
        flush=True,
    )


asyncio.run(main())
